session/session.go

```diff
@@ -1264,6 +1269,41 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
 	return recordSet, nil
 }
 
+func (s *session) validateStatementReadOnlyInStaleness(stmtNode ast.StmtNode) error {
+	vars := s.GetSessionVars()
+	if !vars.TxnCtx.IsStaleness {
+		return nil
+	}
+	// For the case executing:
+	// `START TRANSACTION READ ONLY WITH TIMESTAMP BOUND ...`
+	// `...`
+	// `commit`
+	_, ok := stmtNode.(*ast.CommitStmt)
+	if ok {
+		return nil
+	}
+	// For the case executing:
+	// `START TRANSACTION READ ONLY WITH TIMESTAMP BOUND ...`
+	// `...`
+	// `rollback`
+	_, ok = stmtNode.(*ast.RollbackStmt)
+	if ok {
+		return nil
+	}
+	// For the case executing:
+	// `START TRANSACTION READ ONLY WITH TIMESTAMP BOUND ...`
+	// `...`
+	// `begin` or `START TRANSACTION READ ONLY WITH TIMESTAMP BOUND ...`
+	_, ok = stmtNode.(*ast.BeginStmt)
+	if ok {
+		return nil
+	}
+	if !planner.IsReadOnly(stmtNode, vars) {
```

Comment:
I think some other statement types that not considered in `IsReadOnly` should also be considered. `IsReadOnly` was used for the optimizer and transaction retrial mechanism. However, in your case, those statements not in `IsReadOnly` can't be executed.
For example, DDL / Analyze end the current transaction, so they can be executed.
For these situations, you must check if they will be executed in `ExecuteStmt`. If they do, you should also allow them to be executed.

Reply:
updated.